### PR TITLE
Add support for DarwinArchs when assembling macOS App.framework

### PR DESCRIPTION
**Pull request by knopp**
https://github.com/flutter/flutter/issues/81205

This can be used to specify target architectures and produce fat binaries.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 1]
```diff
@@ -103,6 +103,15 @@ enum HostArtifact {
   pubExecutable,
 }
 
+// TODO(knopp): Remove once darwin artifacts are moved out of darwin-x64
+String _platformDirName(TargetPlatform platform) {
+  if (platform == TargetPlatform.darwin) {
+    return 'darwin-x64';
+  } else {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/artifacts.dart:110]: nit: no else needed since its a return above:

```dart
if (platform == TargetPlatform.darwin) {
  return 'darwin-x64';
}
return getNameForTargetPlatform(platform);
```

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart [hunk 2]
```diff
@@ -202,19 +200,51 @@ class CompileMacOSFramework extends Target {
...
-    if (result != 0) {
-      throw Exception('gen shapshot failed.');
+
+    final List<Future<int>> pending = <Future<int>>[];
+    for (final DarwinArch darwinArch in darwinArchs) {
+      if (codeSizeDirectory != null) {
+        final File codeSizeFile = environment.fileSystem
+          .directory(codeSizeDirectory)
+          .childFile('snapshot.${getNameForDarwinArch(darwinArch)}.json');
+        extraGenSnapshotOptions.add('--write-v8-snapshot-profile-to=${codeSizeFile.path}');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/macos.dart:210]: this seems duplicated

**Review by jonahwilliams — COMMENTED** [review]
We should have some example unit tests for the iOS lipo you could copy from for the macOS example

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart (hunk 2 above)
Comment (knopp) [reply packages/flutter_tools/lib/src/build_system/targets/macos.dart]: Good catch!

**Comment by knopp** [comment]
I'll look at the tests for lipo. Also debug build only produces x86_64 framework, I'll fix that too.

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 3]
```diff
@@ -514,7 +522,7 @@ class CachedArtifacts implements Artifacts {
         // TODO(jonahwilliams): remove once debug desktop artifacts are uploaded
         // under a separate directory from the host artifacts.
         // https://github.com/flutter/flutter/issues/38935
-        String platformDirName = getNameForTargetPlatform(platform);
+        String platformDirName = _platformDirName(platform);
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/artifacts.dart:525]: I realize you probably named `_platformDirName` off these locals, but you can name `_platformDirName` something more descriptive?  Maybe `_enginePlatformDirectoryName`?

File: packages/flutter_tools/lib/src/base/build.dart [hunk 4]
```diff
@@ -311,7 +311,7 @@ class AOTSnapshotter {
       TargetPlatform.android_arm64,
       TargetPlatform.android_x64,
       TargetPlatform.ios,
-      TargetPlatform.darwin_x64,
+      TargetPlatform.darwin,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/base/build.dart:314]: I agree with you in https://github.com/flutter/flutter/issues/81205#issuecomment-827131666
> It does feel a bit off that there is `TargetPlatform.ios` and `--define=IosArchs=...` and then `TargetPlatform.darwin`. But iOS _is_ darwin. It really should be `TargetPlatform.macOS` and `--define=MacOSArchs=...`.

Just glancing at this I wouldn't be sure which to use (the confusion predates this PR).  This also gets hairier when we start targeting iOS apps as native ARM mac apps for the unfortunately named "Mac Designed for iPad" https://github.com/flutter/flutter/issues/66184.

I'd be supportive if you wanted to rename this `macos` instead of `darwin`.

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart [hunk 5]
```diff
@@ -177,21 +186,19 @@ class CompileMacOSFramework extends Target {
...
-        .directory(codeSizeDirectory)
-        .childFile('snapshot.${getNameForDarwinArch(DarwinArch.x86_64)}.json');
-      extraGenSnapshotOptions.add('--write-v8-snapshot-profile-to=${codeSizeFile.path}');
-      final File precompilerTraceFile = environment.fileSystem
-        .directory(codeSizeDirectory)
-        .childFile('trace.${getNameForDarwinArch(DarwinArch.x86_64)}.json');
-      extraGenSnapshotOptions.add('--write-v8-snapshot-profile-to=${codeSizeFile.path}');
-      extraGenSnapshotOptions.add('--trace-precompiler-to=${precompilerTraceFile.path}');
+    final TargetPlatform targetPlatform = getTargetPlatformForName(environment.defines[kTargetPlatform]);
+    final List<DarwinArch> darwinArchs = environment.defines[kDarwinArchs]
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/build_system/targets/macos.dart:195]: Can you add a test for the multi arch splitting?

File: packages/flutter_tools/lib/src/base/build.dart (hunk 4 above)
Comment (knopp) [reply packages/flutter_tools/lib/src/base/build.dart:314]: > I'd be supportive if you wanted to rename this macos instead of darwin.

I'll keep that in mind, but I'm not sure I'd want to do it in this PR.

File: packages/flutter_tools/lib/src/artifacts.dart (hunk 3 above)
Comment (knopp) [reply packages/flutter_tools/lib/src/artifacts.dart]: Sure. `_enginePlatformDirectoryName` sounds good.

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart (hunk 5 above)
Comment (knopp) [reply packages/flutter_tools/lib/src/build_system/targets/macos.dart:247]: Done.

**Comment by knopp** [comment]
I added lipo tests for both framework thinning and universal binary.

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 6]
```diff
@@ -103,6 +103,14 @@ enum HostArtifact {
   pubExecutable,
 }
 
+// TODO(knopp): Remove once darwin artifacts are moved out of darwin-x64
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/artifacts.dart:106]: I don't think we're going to move darwin artifacts out of darwin-x64. We'll still have different engine builds for darwin-x64 and darwin-arm - the target platform is best though of as a convenience for how the engine builds think of targets that was incorrectly extended to how our users interact with the platforms.

Comment (knopp) [reply packages/flutter_tools/lib/src/artifacts.dart:106]: Why do you need different engine builds? Can you just have universal FlutterMacOS.framework (just like you have universal Flutter.framework for ios)? The PR already implements framework thinning. 

I assumed that all the artifacts would be distributed as universal in the end.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/artifacts.dart:106]: No better answer than "that is how the engine builds are configured currently".

Comment (knopp) [reply packages/flutter_tools/lib/src/artifacts.dart:106]: Well, sure, but I assumed that at some point there's going to be work done to build the engine for darwin-x64, darwin-arm64 and lipo them together. Isn't that the whole purpose of not having separate `TargetPlatform.darwin_x64` and `.darwin_arm64`?

Somehow this is done for iOS, why can't it be done for macOS? (plus flutter_tester, font-subset and gen_snapshot).

`CachedArtifacts` doesn't know about `DarwinArchs` so I don't quite see how it would choose the right folder if you want to distribute darwin-arm64 build separately.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/artifacts.dart:106]: You could certainly do that work, but there is no point in changing the cache structure in the tool before that is done

Comment (knopp) [reply packages/flutter_tools/lib/src/artifacts.dart]: I didn't assume this would happen before having universal artifacts. There's no point in doing that. I'll clarify that in the TODO comment.

Comment (jmagman) [reply packages/flutter_tools/lib/src/artifacts.dart]: Quasi related for distribution but I suggested combining the macOS framework into the iOS Flutter.xcframework https://github.com/flutter/flutter/issues/70413 though considering how unfortunately different the two platform projects look, and how manual it was to get the right framework slice out of the iOS xcframework, I'm not now sure if that will be very valuable.  In any case, we should distribute a universal macOS framework.  But no one has worked on that yet.

Comment (knopp) [reply packages/flutter_tools/lib/src/artifacts.dart]: Well, I have no clue how flutter builds work on server side so I'm not sure I can help with that :) However with recent changes landed in buildroot and engine it is now possible to cross compile flutter for darwin-arm64. But it requires manually editing `build.gn`. I think the next step would be to add darwin cross compilation to `flutter/tools/gn`. I'll open an issue for it.

**Review by jonahwilliams — APPROVED** [review]
LGTM if LG to @jmagman

**Review by jmagman — APPROVED** [review]
LGTM, thanks @knopp!

**Comment by zanderso** [comment]
@knopp @jonahwilliams @jmagman I'm having a bit of trouble following the discussion and determining what the missing pieces are. This PR seems to be making some assumptions about how the arm64 artifacts will be supplied and consumed. Can some ELI5 what those are for me?

**Comment by jonahwilliams** [comment]
So IIRC the idea was that we'd change how we were handling architecture - instead of having platform + arch in one enum this changes macos to be more like ios where the platform and arch are separate flags. I don't remember if any additional work was done to support arm64 builds.

**Comment by jmagman** [comment]
Exactly, instead of macOS platform being specified `darwin_x64` instead it's `darwin` + either `x86_64` or `arm64`.  It was mostly refactoring for the sake of the enums.

@zanderso I can do the outstanding in the tool to stop using Rosetta once there's an arm FlutterMacOS.framework, if that's what you're exploring.  I already had to do it for iOS, and it'll be easier with macOS since it's the same artifact paths, but with an extra architecture lipo'd in.

**Comment by knopp** [comment]
> wing the discussion and determining what the missing pieces are. This PR seems to be making some assumptions about how the arm64 artifacts will be supplied and consumed. Can someone ELI5 what those are for me? (e.g. libflutter.so)

This is similar to how iOS work. Flutter tool can produce fat binaries for `App.framework` and strip unneeded architectures from `FlutterMacOS.framework`. IIRC what's left to do is mostly

- Produce fat binary for `FlutterMacOS.framework`. We can already cross compile the framework for ARM64, just need script to lipo the framework and other desktop artifacts and add that to [Luci](https://github.com/flutter/flutter/issues/81488).
- Modify the paths in tool to download and use universal framework
- Modify the tool to stop using rosetta (don't launch build through `arch`)

**Comment by knopp** [comment]
> Exactly, instead of macOS platform being specified darwin_x64 instead it's darwin + either x86_64 or arm64. It was mostly refactoring for the sake of the enums.

Actually, it can be both x86_64 and arm64. Tool should build snapshot for all architectures and then lipo them together.

**Comment by zanderso** [comment]
How about `gen_snapshot`? I think the `gen_snapshot_arm64` from the ios artifacts will work. Can the desktop build also use that?

**Comment by jmagman** [comment]
> > Exactly, instead of macOS platform being specified darwin_x64 instead it's darwin + either x86_64 or arm64. It was mostly refactoring for the sake of the enums.
> 
> Actually, it can be both x86_64 and arm64. Tool should build snapshot for all architectures and then lipo them together.

Yes, I meant the targeted platform can be described as darwin+x86_64 or darwin+arm64.

**Comment by jmagman** [comment]
> How about `gen_snapshot`? I think the `gen_snapshot_arm64` from the ios artifacts will work. Can the desktop build also use that?

If the same `gen_snapshot_arm64` works (not sure if it does) then we'd also need to either ship it in the macos artifacts, or update the tool to always require the iOS artifacts to be downloaded.  Either way, need tooling work to detect the current arch and choose the appropriate `gen_snapshot` if they can't be lipo'd together.

**Comment by knopp** [comment]
> > How about `gen_snapshot`? I think the `gen_snapshot_arm64` from the ios artifacts will work. Can the desktop build also use that?
> 
> If the same `gen_snapshot_arm64` works (not sure if it does) then we'd also need to either ship it in the macos artifacts, or update the tool to always require the iOS artifacts to be downloaded. Either way, need tooling work to detect the current arch and choose the appropriate `gen_snapshot` if they can't be lipo'd together.

Why couldn't `gen_snapshot` be lipod together?
